ts/client/services/events/depositsFinalised.ts

```diff
@@ -0,0 +1,103 @@
+import { BigNumber, Event } from "ethers";
+import { Rollup } from "../../../../types/ethers-contracts";
+import { FeeReceivers } from "../../config";
+import { ContractEventSyncer } from "./contractEventSyncer";
+
+export type PendingDeposit = {
+    tokenID: BigNumber;
+    depositID: BigNumber;
+};
+
+/**
+ * Syncs DepositsFinalised events from the Rollup contract
+ * Currently only used by feeReceivers script.
+ */
+export class DepositsFinalisedEventSyncer extends ContractEventSyncer {
+    public readonly feeReceivers: FeeReceivers;
+    private readonly subtreeIDToPendingDeposits: Record<
+        string,
+        PendingDeposit[]
+    >;
+
+    constructor(
+        rollup: Rollup,
+        private readonly maxDepositSubtreeDepth: number
+    ) {
+        super(rollup, rollup.filters.DepositsFinalised());
+        this.eventListener = this.depositsFinalisedListener;
+
+        this.feeReceivers = [];
+        this.subtreeIDToPendingDeposits = {};
+    }
+
+    public async initialSync(
+        _startBlock: number,
+        _endBlock: number
+    ): Promise<void> {
+        throw new Error(
+            "DepositsFinalisedEventSyncer: initialSync not implemented"
+        );
+    }
+
+    public getPendingDepositsCount(): number {
+        return Object.keys(this.subtreeIDToPendingDeposits).reduce(
+            (acc, k) => acc + this.subtreeIDToPendingDeposits[k].length,
+            0
+        );
+    }
+
+    public addPendingDeposit(
+        subtreeID: BigNumber,
+        pendingDeposit: PendingDeposit
+    ) {
+        const subIDStr = subtreeID.toString();
+        if (!this.subtreeIDToPendingDeposits[subIDStr]) {
+            this.subtreeIDToPendingDeposits[subIDStr] = [];
+        }
+
+        this.subtreeIDToPendingDeposits[subIDStr].push(pendingDeposit);
+    }
+
+    private handleDepositsFinalised(event: Event) {
+        if (!event.args) {
+            throw new Error("missing event args");
+        }
+        const { subtreeID, pathToSubTree } = event.args;
+
+        const subIDStr = subtreeID.toString();
+        console.log(`subtree #${subIDStr} submitted`);
+
+        const pendingDeposits = this.subtreeIDToPendingDeposits[subIDStr];
+        if (!pendingDeposits) {
+            throw new Error(
+                `pending deposits not found for subtreeID ${subIDStr}`
+            );
+        }
+        delete this.subtreeIDToPendingDeposits[subIDStr];
+
+        for (const pd of pendingDeposits) {
+            const stateID = pathToSubTree
+                .mul(2 ** this.maxDepositSubtreeDepth)
+                .add(pd.depositID)
+                .sub(1);
```

Comment:
I just realized that we had to sub(1) here because the `depositID` starts from 1.
We could consider an alternative that starts depositID from 0. It requires another deposit contract fix but we can get consistency for reasoning leaf indices.

Reply:
+1 to that

Reply:
https://github.com/thehubbleproject/hubble-contracts/issues/642